python/src/iceberg/transforms.py

```diff
@@ -0,0 +1,446 @@
+
+    def to_human_string(self, value) -> str:
+        if value is None:
+            return "null"
+        return self._to_human_string(value)
+
+    def dedup_name(self) -> str:
+        return self._transform_string
+
+
+class Bucket(Transform):
+    _MAX_32_BITS_INT = 2147483647
+    _FUNCTIONS_MAP = {  # [0] is hash function and [1] is can_transform check function
+        DateType: (
+            lambda v: mmh3.hash(struct.pack("q", v)),
+            lambda t: t in [IntegerType, DateType],
+        ),
+        IntegerType: (
+            lambda v: mmh3.hash(struct.pack("q", v)),
+            lambda t: t in [IntegerType, DateType],
+        ),
+        TimeType: (
+            lambda v: mmh3.hash(struct.pack("q", v)),
+            lambda t: t in {LongType, TimeType, TimestampType, TimestamptzType},
+        ),
+        TimestampType: (
+            lambda v: mmh3.hash(struct.pack("q", v)),
+            lambda t: t in {LongType, TimeType, TimestampType, TimestamptzType},
+        ),
+        TimestamptzType: (
+            lambda v: mmh3.hash(struct.pack("q", v)),
+            lambda t: t in {LongType, TimeType, TimestampType, TimestamptzType},
+        ),
+        LongType: (
+            lambda v: mmh3.hash(struct.pack("q", v)),
+            lambda t: t in [LongType, TimeType, TimestampType, TimestamptzType],
+        ),
+        StringType: (
+            lambda v: mmh3.hash(v),
+            lambda t: t == StringType,
+        ),
+        BinaryType: (
+            lambda v: mmh3.hash(v),
+            lambda t: t == BinaryType,
+        ),
+        UUIDType: (
+            lambda v: mmh3.hash(
+                struct.pack(
+                    ">QQ",
+                    (v.int >> 64) & 0xFFFFFFFFFFFFFFFF,
+                    v.int & 0xFFFFFFFFFFFFFFFF,
+                )
+            ),
+            lambda t: t == UUIDType,
+        ),
+        # bucketing by Float/Double is not allowed by the spec, but they have hash implementation
+        FloatType: (
+            lambda v: mmh3.hash(struct.pack("d", v)),
+            lambda t: t == FloatType,
+        ),
+        DoubleType: (
+            lambda v: mmh3.hash(struct.pack("d", v)),
+            lambda t: t == DoubleType,
+        ),
+    }
+
+    def __init__(self, source_type: Type, num_buckets: int):
+        if (
+            source_type not in Bucket._FUNCTIONS_MAP
+            and not isinstance(source_type, FixedType)
+            and not isinstance(source_type, DecimalType)
+        ):
+            raise ValueError(f"Cannot bucket by type: {source_type}")
+
+        super().__init__(
+            f"bucket[{num_buckets}]",
+            f"transforms.bucket(source_type={repr(source_type)}, num_buckets={num_buckets})",
+        )
+        self._type = source_type
+        self._num_buckets = num_buckets
+
+        if isinstance(self._type, FixedType):
+            self._hash_func = lambda v: mmh3.hash(v)
+            self._can_transform = lambda t: isinstance(t, FixedType)
```

Comment:
It's hard to read this class having some of these inline and others in the map. Can you move all of these implementations to the same place?

Reply:
SG, I moved them all into methods in transform_utils.